timelock-impl/src/main/java/com/palantir/atlasdb/timelock/auth/AuthManager.java

```diff
@@ -0,0 +1,125 @@
+
+import org.immutables.value.Value;
+
+import com.palantir.atlasdb.http.PollingRefreshable;
+import com.palantir.atlasdb.timelock.auth.api.AuthenticatedClient;
+import com.palantir.atlasdb.timelock.auth.api.Authenticator;
+import com.palantir.atlasdb.timelock.auth.api.Authorizer;
+import com.palantir.atlasdb.timelock.auth.api.BCryptedSecret;
+import com.palantir.atlasdb.timelock.auth.api.Credentials;
+import com.palantir.atlasdb.timelock.auth.api.Password;
+import com.palantir.atlasdb.timelock.auth.api.Privileges;
+import com.palantir.atlasdb.timelock.auth.config.PrivilegesConfiguration;
+import com.palantir.atlasdb.timelock.auth.config.TimelockAuthConfiguration;
+import com.palantir.atlasdb.timelock.auth.impl.AuthRequirement;
+import com.palantir.atlasdb.timelock.auth.impl.CachingAuthenticator;
+import com.palantir.atlasdb.timelock.auth.impl.SimpleAuthorizer;
+import com.palantir.conjure.java.ext.refresh.Refreshable;
+import com.palantir.lock.TimelockNamespace;
+
+public class AuthManager implements AutoCloseable {
+    private final Refreshable<TimelockAuthConfiguration> authConfigurationRefreshable;
+    private final Runnable closingCallback;
+    private volatile AuthServices authServices;
+
+    AuthManager(Refreshable<TimelockAuthConfiguration> authConfigurationRefreshable,
+            Runnable closingCallback,
+            AuthServices authServices) {
+        this.authConfigurationRefreshable = authConfigurationRefreshable;
+        this.closingCallback = closingCallback;
+        this.authServices = authServices;
+    }
+
+    public static AuthManager of(Supplier<TimelockAuthConfiguration> timelockAuthConfigurationSupplier) {
+        PollingRefreshable<TimelockAuthConfiguration> configPollingRefreshable = PollingRefreshable
+                .create(timelockAuthConfigurationSupplier);
+
+        AuthServices authServices = getAuthServices(timelockAuthConfigurationSupplier.get());
+
+        return new AuthManager(
+                configPollingRefreshable.getRefreshable(),
+                configPollingRefreshable::close,
+                authServices);
+    }
+
+    public void checkAuthorized(String clientId, Password password, TimelockNamespace timelockNamespace) {
+        update();
+
+        if (!authServices.useAuth()) {
+            return;
+        }
+
+        Optional<AuthenticatedClient> client = authServices.authenticator().authenticate(clientId, password);
+        if (!client.isPresent() || !authServices.authorizer().isAuthorized(client.get(), timelockNamespace)) {
+            throw new ForbiddenException();
+        }
+    }
+
+    @Override
+    public void close() {
+        closingCallback.run();
+    }
+
+    private void update() {
+        authConfigurationRefreshable.getAndClear().ifPresent(this::updateInternal);
+    }
+
+    private synchronized void updateInternal(TimelockAuthConfiguration authConfiguration) {
+        authServices = getAuthServices(authConfiguration);
```

Comment:
doesn't need to be synchronized anymore, also needs to set and return `authServices`

Reply:
might be convenient to be use `AtomicReference` here. `getAndSet`